rfcs/2021-11-12-9811-vrl-vm.md

```diff
@@ -0,0 +1,164 @@
+stored in the `values` `Vec` that is found in position 12 onto the stack.
+
+- values
+
+A list of constant values found in the program. Since the bytecode only
+contains integers any actual values must be stored here. This also allows
+literals to be deduped.
+
+- targets
+
+A list of paths used in the program, similar to `values`.
+
+- stack
+
+The Vm is a stack based Vm. Every expression that is evaluated pushes the
+result on the stack. Every operation pulls the values it uses from the stack.
+
+- ip
+
+The instruction pointer points to the next instruction to evaluate.
+
+With each node of the AST compiled down to just a few bytes and all
+instructions held in contiguous memory evaluation of the program should be able
+to take full advantage of the CPU cache which should result in much faster
+execution.
+
+
+#### Calling functions
+
+Calling functions in the stdlib will be a case of evaluating each parameter
+with the results pushed onto the stack.
+
+Since Vrl allows for named parameters and optional parameters, the compiler
+will need to ensure the bytecode evaluates each parameter in a specific order
+(likely the order declared in the function). Bytecode will need to be emmitted
+for parameters that are not specified in the Vrl script to add a default value
+to the stack.
+
+
+### Optimization
+
+With the code as a single dimension array of Bytecode, it could be possible to
+scan the code for patterns and reorganise the Bytecode so it can run in a more
+optimal way.
+
+A lot more thought and research needs to go into this before we can consider
+implementing these changes.
+
+## Drawbacks
+
+Downsides to using a Vm:
+
+- The code is a bit more complex. With an AST that is walked it is fairly apparent
+  what the code will be doing at any point. With a Vm, this is not the case, it
+  is harder to look at the instructions in the Vm and follow back to what part
+  of the Vrl code is being evaluated. We will need to write some extensive
+  debugging tools to allow for decent introspection into the Vm.
+
+  However, VRL is a very simple language, which does also mean the VM will be
+  simple.
+
+- We lose some safety that we get from the Rust compiler. There will need
+  to be significant fuzz testing to ensure that the code runs correctly under
+  all circumstances. It should be noted however that the entire VM will not
+  require any _unsafe_ code.
+
+- Currently each stdlib function is responsible for evaluating their own
+  parameters. This allows parameters to be lazily evaluated. Most likely with a
+  Vm, the parameters will need to be evaluated up front and the stack passed
+  into the function. This could impact performance.
+
+## Prior Art
+
+- Goscript - An implementation of Go using a bytecode Vm,
+  https://github.com/oxfeeefeee/goscript
+
+- CPython - https://github.com/python/cpython
+
+## Plan Of Attack
+
+- [ ] Submit a PR with spike-level code _roughly_ demonstrating the change for
+      the VM.
```

Comment:
This needs to be fleshed out more.

Specifically, I'd like to see the approach we discussed listed here:

1. Introduce/implement the VM itself, without updating any of the expressions. This includes tests (unit/property), documentation/code comments, and an as-safe-as-possible API surrounding the VM operations. 
2. Update one or two expressions to set the tone for future work of updating all the other expressions and functions (if needed)
3. Divvy up the work to update all other expressions(/functions).

Reply:
I think if we create a decent wrapper for Functions number 2 in this list could potentially be a simple `sed` call.